### PR TITLE
Improve CommandCallable. Fixes #465

### LINKED ISSUE #465: Remove Optional from CommandCallable
I think it doesn't really make sense that the usage string is required while the description and help strings are wrapped in an ``Optional``. It should be required that every command has a help message!

I think in this case, it would be better to use simple strings. **That also makes it easier for plugin developers to implement the interface.**

Example how it is right now: https://docs.spongepowered.org/en/plugins/commands.html

What I propose:
```java
String getShortDescription() {
    return "Displays a message to all players";
}
String getHelp() {
    return "Displays a message to all players. It has no color support!";
}
String getUsage() {
    return "<message>";
}
```

### PR DESCRIPTION
**Pull request by boformer**
Remove ``Optional`` from ``CommandCallable``
------------------

I suggest not to use the ``Optional`` return type in the ``CommandCallable`` interface.

The return type is used for the short description and the help message of the command. I think that **every command (and every dispatcher) should have a description and a help message.** Making the messages non-optional encourages the developers to write proper help messages. Help systems can use these messages to create a useful in-game help, which is important for players.

Usually only the Sponge implementation (and third-party help systems) will call the methods of the interface. These systems must check for ``null`` to prevent NPEs, even when the interface uses ``Optional`` (plugin developers are lazy, so they use their IDE to generate a "return null"-method). 

Removing the ``Optional`` return type also makes it easier for plugin developers to implement the interface:

```java
String getShortDescription() {
    return "Displays a message to all players";
}
Message getHelp() {
    return Messages.of("Displays a message to all players. It has color support!");
}
String getUsage() {
    return "<message>";
}
```
Formatted help message
-----------------------

The second change I propose is to use ``Message`` instead of a string for the longer help message.
This allows the use of ``\n`` for multiline messages. It also allows developers to embed hyperlinks to the plugin documentation.

The only flaw I see is that developers can abuse the formatting support (obfuscated, bold, red text).
I added a few recommendations in the javadoc to stop developers from overusing the formatting features.

Add source argument
-----------------------

I also propose to pass a command source to the help methods. This allows the command to return different help messages depending on the permission of the source:

```java
//Teleport command /tp [player] <player>
String getUsage(CommandSource source) {
    if(source.hasPermission("example.tp.other") 
        return "[player] <player>"; //teleport self or other player to target player
    else if(source.hasPermission("example.tp.self") 
        return "<player>"; //only teleport self to target player
}
```

Add better constructor to ``SimpleDispatcher``
------------------
The last change is an optional constructor for ``SimpleDispatcher`` that sets the help messages of the dispatcher. With this change, plugin developers can setup subcommand structures with detailed help messages without creating a custom dispatcher class.

### CODE REVIEW

File: src/main/java/org/spongepowered/api/util/command/CommandCallable.java [hunk 1]
```diff
@@ -69,25 +71,43 @@
...
 
     /**
-     * Get a longer help text about this command.
+     * Get a longer formatted help message about this command.
+     * 
+     * <p>It is recommended to use the default text color and style. Sections 
+     * with text actions should be underlined.</p>
+     * 
+     * <p>Multiline messages can be created by separating the lines with 
+     * {@code \n}</p>
```
Comment (gratimax) [inline src/main/java/org/spongepowered/api/util/command/CommandCallable.java]: missing .

**Comment by gratimax** [comment]
The Optionals are definitely not changing for consistency with the rest of the API, and some of the javadocs are copy-pasted from my PR on the documentation, but otherwise I'm okay with the rest of the changes.

**Comment by boformer** [comment]
@gratimax In that case, let's just make the help texts required (remove the @Nullable).
A command description is like the img alt tag in html and should not be missing.

**Comment by gratimax** [comment]
Hrmmm, I retract the copy-paste comment. I was confused because the javadocs are talking about similar things compared to the Docs PR, yet don't flow very well.

**Comment by Minecrell** [comment]
I agree with @gratimax here, we're not going to use `@Nullable` just for _some_ interfaces/methods as this will just cause confusion at the end. Why not just return an empty string and `Texts.EMPTY` instead of null?

**Comment by boformer** [comment]
@Minecrell sounds good. I will remove the annotation.

**Comment by boformer** [comment]
@gratimax I think it is quite important to specify the recommended format in the javadocs. This line of text will improve the consistency of the help messages.

**Comment by boformer** [comment]
I think this is ready.

**Comment by me4502** [comment]
My only question, is why can't the short description be a Message as well?

**Comment by boformer** [comment]
> why can't the short description be a Message as well?

@me4502 The short description is displayed in the command list, next to the command.
There is no reason why a plugin developer should be able to user colors, text styles, click actions or multiple lines in this description.

```
--- Command Help [Page 1/7] ---
/ban - Ban players
/broadcast - {This is the short description}
/kill - Kill players or entities
/msg - Send a message to a single player
/tp - Teleport to a player or location
..
Usage: /help [page] or /help [command]
```

**Comment by me4502** [comment]
Fair enough

File: src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java [hunk 2]
```diff
@@ -54,6 +57,27 @@
...
+    }
+    
+    /**
+     * Creates a new dispatcher with a description and a help message.
+     *
+     * @param shortDescription A short one-line description
+     * @param help A formatted help message
+     */
+    public SimpleDispatcher(String shortDescription, Message help) {
+        this.shortDescription = shortDescription;
```
Comment (Minecrell) [inline src/main/java/org/spongepowered/api/util/command/dispatcher/SimpleDispatcher.java]: Null checks would be good here

**Comment by Zidane** [comment]
@SpongePowered/developers I want feedback on this.

**Comment by me4502** [comment]
I'm all For this

**Comment by Aaron1011** [comment]
@boformer: You should update this to the 'Text API 2' (`Message` was renamed to `Text`, among other things).

Otherwise, this looks great!

**Comment by boformer** [comment]
@Aaron1011 updated to Text API v2
@Minecrell added null checks.

I think it is ready for merge.

**Comment by me4502** [comment]
I just read through the PR, it all looks fine.

**Comment by Zidane** [comment]
I had to manually commit this. In the future, please don't issue a merge commit in a PR... :/.

**Pull request closed without merging**